<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Introduction · Whale.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link href="assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>Whale.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li class="current"><a class="toctext" href>Introduction</a><ul class="internal"><li><a class="toctext" href="#Installation-1">Installation</a></li><li><a class="toctext" href="#Data-preparation-1">Data preparation</a></li><li><a class="toctext" href="#Quick-start-1">Quick start</a></li><li><a class="toctext" href="#References-1">References</a></li></ul></li><li><a class="toctext" href="manual/">Manual</a></li><li><a class="toctext" href="api/">API</a></li></ul></nav><article id="docs"><header><nav><ul><li><a href>Introduction</a></li></ul><a class="edit-page" href="https://github.com/arzwa/Whale.jl/blob/master/docs/src/index.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Introduction</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Introduction-1" href="#Introduction-1">Introduction</a></h1><div class="admonition warning"><div class="admonition-title">Warning</div><div class="admonition-text"><p>The latest Whale version is a thorough rewrite of the Whale library, and is still work in progress. For the version as used in Zwaenepoel &amp; Van de Peer (2019), refer to <a href="https://github.com/arzwa/Whale.jl/releases/tag/v0.2">this release (v0.2)</a>. Nevertheless, the current version should be safe to use and is more efficient and convenient (if you know a bit of julia).</p></div></div><p>Whale provides tools for genome-wide <strong>amalgamated likelihood estimation (ALE) under a DL+WGD model</strong>, which is an approach to infer reconciled gene trees and parameters of a model of gene family evolution given a known species tree.</p><p>The ALE approach takes into account uncertainty in the gene tree topology by marginalizing over all tree topologies that can be amalgamated from the so-called <em>conditional clade distribution</em> (CCD). This CCD can be constructed from a sample of the posterior distribution of tree topologies (which can be obtained using any standard software for Bayesian phylogenetics).</p><p>More specifically, this library can be used to</p><ul><li>Statistically test the absence or presence of hypothetical whole-genome duplication (WGD) events in a species phylogeny</li><li>Infer lineage-specific gene duplication and loss rates for a species phylogeny</li><li>Infer high-quality (reconciled) gene trees given a known species tree <a href="https://academic.oup.com/sysbio/article/64/1/e42/1634124">cf. Szöllősi <em>et al.</em></a></li><li>All of the above at once</li></ul><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>This library implements the DL and DL+WGD models. It does not implement models of gene family evolution that take into account horizontal gene transfer, incomplete lineage sorting or gene conversion.</p></div></div><h2><a class="nav-anchor" id="Installation-1" href="#Installation-1">Installation</a></h2><p>You will need julia-1.x. Fire up a julia REPL by typing <code>julia</code> at the command line enter the package manager interface by typing <code>]</code>, and run the command</p><pre><code class="language-julia-repl">(v1.1) pkg&gt; add https://github.com/arzwa/PhyloTrees.jl
(v1.1) pkg&gt; add https://github.com/arzwa/ConsensusTrees.jl
(v1.1) pkg&gt; add https://github.com/arzwa/Whale.jl</code></pre><h2><a class="nav-anchor" id="Data-preparation-1" href="#Data-preparation-1">Data preparation</a></h2><p>To perform analyses with Whale, you will need  </p><ol><li>An <strong>ultrametric species tree</strong>, with ideally branch lengths in geological time (since this allows straightforward interpretation of parameter estimates.)</li><li>A bunch of ALE files, which summarize the <strong>conditional clade distributions</strong> (CCDs) for the same bunch of gene families. These can be obtained from a sample of the posterior distribution of gene trees using the <a href="https://github.com/ssolo/ALE"><code>ALEobserve</code></a> tool. A pipeline to obtain these from a set of gene family protein fasta files is available at <a href="https://github.com/arzwa/whaleprep">github</a>.</li></ol><h2><a class="nav-anchor" id="Quick-start-1" href="#Quick-start-1">Quick start</a></h2><p>If you&#39;re not familiar with <code>julia</code>, and you simply want to run analyses as performed for instance in <a href="https://academic.oup.com/mbe/advance-article/doi/10.1093/molbev/msz088/5475503">Zwaenepoel &amp; Van de Peer (2019)</a> the following scripts will be helpful. If you want to get a more detailed view of the library, please consult the <a href="manual/#Manual-1">Manual</a>.</p><h3><a class="nav-anchor" id="Maximum-likelihood-estimation-1" href="#Maximum-likelihood-estimation-1">Maximum likelihood estimation</a></h3><p>I assume that your tree is called <code>tree_file.nw</code> and your <code>.ale</code> files are in a directory <code>ccd_dir</code>. For maximum likelihood estimation with a constant rates model, the following script should work</p><pre><code class="language-julia">@everywhere using Whale

# data and config
st = SlicedTree(&quot;tree_file.nw&quot;)
ccd = read_ale(&quot;ccd_dir&quot;, st)
constant = true  # set to false for branch-wise rates

# inference
constant ? set_constantrates!(st) : nothing
w = WhaleModel(st)
out = mle(w, ccd)</code></pre><p>Save this script to a file (say <code>whale-mle.jl</code>). To run the inference using 16 processors for instance, run <code>julia -p 16 whale-mle.jl</code>. Do not forget the <code>@everywhere</code> before the <code>using</code> statement, as this will load the Whale library on all available processors.</p><p>To add WGDs, a <code>wgd_conf</code> argument should be provide to <code>SLicedTree</code>, please see the docs for <a href="api/#Whale.SlicedTree"><code>SlicedTree</code></a>. See the manual section on <a href="manual/#Rate-indices-1">Rate indices</a> on how to specify local-clock models.</p><div class="admonition warning"><div class="admonition-title">Warning</div><div class="admonition-text"><p>Depending on the time scale and data set at hand, you may need to tweak the initial values of the <code>WhaleModel</code> to prevent Numerical issues!</p></div></div><div class="admonition warning"><div class="admonition-title">Warning</div><div class="admonition-text"><p>ML estimation with branch-wise rates may result in poor convergence for large species trees.</p></div></div><h3><a class="nav-anchor" id="Bayesian-inference-with-MCMC-1" href="#Bayesian-inference-with-MCMC-1">Bayesian inference with MCMC</a></h3><p>A similar script for Bayesian inference using MCMC looks like (here with a hypothetical WGD configuration)</p><pre><code class="language-julia">@everywhere using Whale

# data and config
wgd_conf = Dict(
    &quot;wgd1&quot; =&gt; (&quot;taxon1,taxon2&quot;, 1.2),  # (LCA, time BP)
    &quot;wgd2&quot; =&gt; (&quot;taxon1,taxon5&quot;, 3.2)
)
st = SlicedTree(&quot;tree_file.nw&quot;, wgd_conf)
ccd = read_ale(&quot;ccd_dir&quot;, st)
model = IRModel(st)      # independent log-normal rates
# model = GBMModel(st)   # autocorrelated rates
n = 11000

w = WhaleChain(st, model)
chain = mcmc!(w, ccd, n, show_every=10)</code></pre><p>again, saving this as <code>whale-bay.jl</code> and running this with <code>julia -p &lt;nCPU&gt; whale-bay.jl</code> will start the MCMC with the likelihood evaluation performed on <code>nCPU</code> cores.</p><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><p>You might want to save the MCMC simulation to a file. In the above example one can either save the <code>chain</code> variable to a file using <a href="https://github.com/JuliaIO/JLD.jl"><code>JLD</code></a>, save the data frame in the field <code>chain.df</code> using <code>CSV.write</code> or alternatively one can just set the <code>show_every</code> argument to the desired thinning level and  redirect <code>stdout</code> to a file while running the MCMC.</p></div></div><div class="admonition warning"><div class="admonition-title">Warning</div><div class="admonition-text"><p>Please take you time to understand the hierarchical model used in Whale and to modify the prior distributions to suit your data set! See the <a href="manual/#Bayesian-inference-1">Bayesian inference</a> section of the manual.</p></div></div><h2><a class="nav-anchor" id="References-1" href="#References-1">References</a></h2><p><code>Whale.jl</code> was developed by Arthur Zwaenepoel at the VIB-UGent center for plant systems biology (bioinformatics &amp; evolutionary genomics group). If you use Whale, please cite:</p><blockquote><p>Zwaenepoel, A. and Van de Peer, Y., 2019. Inference of Ancient Whole-Genome Duplications and the Evolution of Gene Duplication and Loss Rates. <em>Molecular biology and evolution</em>, 36(7), pp.1384-1404.</p></blockquote><p>The methods in Whale are heavily inspired by previous work done by other researchers. If you use Whale, consider citing the following two particularly important studies:</p><blockquote><p>[ALE] Szöllősi, G.J., Rosikiewicz, W., Boussau, B., Tannier, E. and Daubin, V., 2013. Efficient exploration of the space of reconciled gene trees. <em>Systematic biology</em>, 62(6), pp.901-912.</p></blockquote><blockquote><p>[DL+WGD model] Rabier, C.E., Ta, T. and Ané, C., 2013. Detecting and locating whole genome duplications on a phylogeny: a probabilistic approach. <em>Molecular biology and evolution</em>, 31(3), pp.750-762.</p></blockquote><footer><hr/><a class="next" href="manual/"><span class="direction">Next</span><span class="title">Manual</span></a></footer></article></body></html>
